### DISCUSSION TITLE
Using Next.JS Middleware to rewrite to a different static shell

### DISCUSSION THREAD

**Discussion by bitttttten** [vercel/community#6584] [Help]
### Summary

We have an app with a few variants depending on the user status. An example is one route with two variants, one for logged in users and one for logged out users.

We want to take advantage of static rendering so I'm wondering if it's possible to use middleware to detect the log in status and then rewrite to a different static shell ala Partial Prerendering although the static shell that is first served is different depending on _some data or context_. 

I'll give an example of the way I imagine it would work: so we have `app/about/page.tsx` although in here we'd like at least two variants. Let's say `app/about/_shells/logged-out-page.tsx` and `app/about/_shells/logged-in-page.tsx` and we'd chose one for the `page.tsx` depending on the state.

The idea is that

1. This would happen at Edge and stay with in the Edge Cache since static shells are kept on the Edge Cache.
2. This would all interact with static content.

### Steps to Reproduce

n/a

**Comment by versecafe** [comment]
@bitttttten you can use a middleware.ts like this
```ts
import { NextRequest, NextResponse } from "next/server";

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    {
      source: "/((?!api|_next/static|_next/image|favicon.ico).*)",
      missing: [
        { type: "header", key: "next-router-prefetch" },
        { type: "header", key: "purpose", value: "prefetch" },
      ],
    },
  ],
};

export function middleware(request: NextRequest) {
  console.log("Pathname: ", request.nextUrl.pathname);
  if (request.nextUrl.pathname == "/about") {
    // Note don't check if bool as all strings are truthy
    const authed: boolean = request.cookies.get("authed") === "true";
    // rewrite masking instead of redirecting
    if (authed) {
      return NextResponse.rewrite(
        new URL("/about/_shells/logged-in", request.url),
      );
    } else {
      return NextResponse.rewrite(
        new URL("/about/_shells/logged-out", request.url),
      );
    }
  }
}
```

which if the auth cookie is set to "true" will send them to the logged in shell else the logged out shell, but the URL will remain `https://yourdomain.com/about` In this example the shells are just normal pages so `app/about/_shells/logged-out/page.tsx` `app/about/_shells/logged-in/page.tsx`

**Reply by bitttttten** [reply]
Interesting okay. What's the missing array for in the config?

**Reply by versecafe** [reply]
missing is to just skip any request that has those headers, you don't need to include it but it avoids middleware invocations on prefetch calls
